File: src/133_clone_graph/CloneGraph.java
import java.util.*;

class Node {
    public int val;
    public List<Node> neighbors;

    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }

    public Node(int val) {
        this.val = val;
        neighbors = new ArrayList<Node>();
    }

    public Node(int val, ArrayList<Node> neighbors) {
        this.val = val;
        this.neighbors = neighbors;
    }
}

public class CloneGraph {
    public Node cloneGraph(Node node) {
        // ...existing code...
        return null;
    }

    // Helper to build a graph from adjacency list
    private static Node buildGraph(int[][] adj) {
        if (adj.length == 0)
            return null;
        Node[] nodes = new Node[adj.length];
        for (int i = 0; i < adj.length; ++i)
            nodes[i] = new Node(i + 1);
        for (int i = 0; i < adj.length; ++i)
            for (int n : adj[i])
                nodes[i].neighbors.add(nodes[n - 1]);
        return nodes[0];
    }

    // Helper to print adjacency list
    private static void printGraph(Node node) {
        if (node == null) {
            System.out.println("[]");
            return;
        }
        Set<Integer> visited = new HashSet<>();
        List<String> res = new ArrayList<>();
        dfsPrint(node, visited, res);
        Collections.sort(res);
        System.out.println(res);
    }

    private static void dfsPrint(Node node, Set<Integer> visited, List<String> res) {
        if (visited.contains(node.val))
            return;
        visited.add(node.val);
        List<Integer> neigh = new ArrayList<>();
        for (Node n : node.neighbors)
            neigh.add(n.val);
        Collections.sort(neigh);
        res.add("(" + node.val + ", " + neigh + ")");
        for (Node n : node.neighbors)
            dfsPrint(n, visited, res);
    }

    public static void main(String[] args) {
        // Test case 1: Empty graph
        Node node = buildGraph(new int[][] {});
        Node cloned = new CloneGraph().cloneGraph(node);
        System.out.println(cloned == null); // true

        // Test case 2: Single node, no neighbors
        node = buildGraph(new int[][] { {} });
        cloned = new CloneGraph().cloneGraph(node);
        printGraph(cloned); // [(1, [])]

        // Test case 3: Two nodes, one edge
        node = buildGraph(new int[][] { { 2 }, { 1 } });
        cloned = new CloneGraph().cloneGraph(node);
        printGraph(cloned); // [(1, [2]), (2, [1])]

        // Test case 4: Four nodes, square
        node = buildGraph(new int[][] { { 2, 4 }, { 1, 3 }, { 2, 4 }, { 1, 3 } });
        cloned = new CloneGraph().cloneGraph(node);
        printGraph(cloned); // [(1, [2, 4]), (2, [1, 3]), (3, [2, 4]), (4, [1, 3])]

        // Test case 5: Three nodes, chain
        node = buildGraph(new int[][] { { 2 }, { 1, 3 }, { 2 } });
        cloned = new CloneGraph().cloneGraph(node);
        printGraph(cloned); // [(1, [2]), (2, [1, 3]), (3, [2])]
    }
}

File: src/133_clone_graph/clone_graph.cpp
#include <algorithm>
#include <functional>
#include <iostream>
#include <set>
#include <vector>
using namespace std;

class Node {
   public:
    int val;
    vector<Node*> neighbors;
    Node() : val(0), neighbors() {}
    Node(int _val) : val(_val), neighbors() {}
    Node(int _val, vector<Node*> _neighbors) : val(_val), neighbors(_neighbors) {}
};

class Solution {
   public:
    Node* cloneGraph(Node* node) {
        if (!node) return nullptr;
        Node* newNode = new Node(node->val);
        for (Node* neighbor : node->neighbors) {
            newNode->neighbors.push_back(cloneGraph(neighbor));
        }
        return newNode;
    }
};

// Helper to build a graph from adjacency list
Node* buildGraph(const vector<vector<int>>& adj) {
    if (adj.empty()) return nullptr;
    vector<Node*> nodes(adj.size(), nullptr);
    for (int i = 0; i < adj.size(); ++i) nodes[i] = new Node(i + 1);
    for (int i = 0; i < adj.size(); ++i)
        for (int n : adj[i]) nodes[i]->neighbors.push_back(nodes[n - 1]);
    return nodes[0];
}
// Helper to print adjacency list
void printGraph(Node* node) {
    if (!node) {
        cout << "[]" << endl;
        return;
    }
    set<int> visited;
    vector<pair<int, vector<int>>> res;
    function<void(Node*)> dfs = [&](Node* n) {
        if (visited.count(n->val)) return;
        visited.insert(n->val);
        vector<int> neigh;
        for (auto x : n->neighbors) neigh.push_back(x->val);
        sort(neigh.begin(), neigh.end());
        res.push_back({n->val, neigh});
        for (auto x : n->neighbors) dfs(x);
    };
    dfs(node);
    sort(res.begin(), res.end());
    for (auto& p : res) {
        cout << "(" << p.first << ", [";
        for (size_t i = 0; i < p.second.size(); ++i) {
            cout << p.second[i];
            if (i + 1 < p.second.size()) cout << ", ";
        }
        cout << "])";
    }
    cout << endl;
}

int main() {
    // Test case 1: Empty graph
    Node* node = buildGraph({});
    Node* cloned = Solution().cloneGraph(node);
    cout << (cloned == nullptr) << endl;  // true

    // Test case 2: Single node, no neighbors
    node = buildGraph({{}});
    cloned = Solution().cloneGraph(node);
    printGraph(cloned);  // (1, [])

    // Test case 3: Two nodes, one edge
    node = buildGraph({{2}, {1}});
    cloned = Solution().cloneGraph(node);
    printGraph(cloned);  // (1, [2])(2, [1])

    // Test case 4: Four nodes, square
    node = buildGraph({{2, 4}, {1, 3}, {2, 4}, {1, 3}});
    cloned = Solution().cloneGraph(node);
    printGraph(cloned);  // (1, [2, 4])(2, [1, 3])(3, [2, 4])(4, [1, 3])

    // Test case 5: Three nodes, chain
    node = buildGraph({{2}, {1, 3}, {2}});
    cloned = Solution().cloneGraph(node);
    printGraph(cloned);  // (1, [2])(2, [1, 3])(3, [2])
    return 0;
}
